src/librustc/mir/transform/dataflow.rs

```diff
@@ -0,0 +1,419 @@
+use mir::repr as mir;
+use mir::cfg::CFG;
+use mir::repr::BasicBlock;
+use rustc_data_structures::bitvec::BitVector;
+
+use mir::transform::lattice::Lattice;
+
+
+pub trait DataflowPass<'tcx> {
+    type Lattice: Lattice;
+    type Rewrite: Rewrite<'tcx, Self::Lattice>;
+    type Transfer: Transfer<'tcx, Self::Lattice>;
+}
+
+pub trait Rewrite<'tcx, L: Lattice> {
+    /// The rewrite function which given a statement optionally produces an alternative graph to be
+    /// placed in place of the original statement.
+    ///
+    /// The 2nd BasicBlock *MUST NOT* have the terminator set.
+    ///
+    /// Correctness precondition:
+    /// * transfer_stmt(statement, fact) == transfer_stmt(rewrite_stmt(statement, fact))
+    /// that is, given some fact `fact` true before both the statement and relacement graph, and
+    /// a fact `fact2` which is true after the statement, the same `fact2` must be true after the
+    /// replacement graph too.
+    fn stmt(&mir::Statement<'tcx>, &L, &mut CFG<'tcx>) -> StatementChange<'tcx>;
+
+    /// The rewrite function which given a terminator optionally produces an alternative graph to
+    /// be placed in place of the original statement.
+    ///
+    /// The 2nd BasicBlock *MUST* have the terminator set.
+    ///
+    /// Correctness precondition:
+    /// * transfer_stmt(terminator, fact) == transfer_stmt(rewrite_term(terminator, fact))
+    /// that is, given some fact `fact` true before both the terminator and relacement graph, and
+    /// a fact `fact2` which is true after the statement, the same `fact2` must be true after the
+    /// replacement graph too.
+    fn term(&mir::Terminator<'tcx>, &L, &mut CFG<'tcx>) -> TerminatorChange<'tcx>;
+}
+
+/// This combinator has the following behaviour:
+///
+/// * Rewrite the node with the first rewriter.
+///   * if the first rewriter replaced the node, 2nd rewriter is used to rewrite the replacement.
+///   * otherwise 2nd rewriter is used to rewrite the original node.
+pub struct RewriteAndThen<'tcx, R1, R2>(::std::marker::PhantomData<(&'tcx (), R1, R2)>);
+impl<'tcx, L, R1, R2> Rewrite<'tcx, L> for RewriteAndThen<'tcx, R1, R2>
+where L: Lattice, R1: Rewrite<'tcx, L>, R2: Rewrite<'tcx, L> {
+    fn stmt(s: &mir::Statement<'tcx>, l: &L, c: &mut CFG<'tcx>) -> StatementChange<'tcx> {
+        let rs = <R1 as Rewrite<L>>::stmt(s, l, c);
+        match rs {
+            StatementChange::None => <R2 as Rewrite<L>>::stmt(s, l, c),
+            StatementChange::Remove => StatementChange::Remove,
+            StatementChange::Statement(ns) =>
+                match <R2 as Rewrite<L>>::stmt(&ns, l, c) {
+                    StatementChange::None => StatementChange::Statement(ns),
+                    x => x
+                },
+            StatementChange::Statements(nss) => {
```

Comment:
> I think that, in the case of multiple statements, just transforming each statement doesn't work - the fact needs to be transferred across each statement.

That’s right, I was thinking about how I would approach it and that’s exactly why I didn’t include a way to replace a statement with a (sub-)graph but replacement with a list of statements fell through the cracks.